scientific_library/tvb/simulator/simulator.py

```diff
@@ -363,6 +363,38 @@ def _loop_monitor_output(self, step, state):
         if any(outputi is not None for outputi in output):
             return output
 
+    def bound_and_clamp(self, state):
+        # If there is a state boundary...
+        if self.integrator.state_variable_boundaries is not None:
+            # ...use the integrator's bound_state
+            self.integrator.bound_state(state)
+        # If there is a state clamping...
+        if self.integrator.clamped_state_variable_values is not None:
+            # ...use the integrator's clamp_state
+            self.integrator.clamp_state(state)
+
+    def _update_and_bound_history(self, history):
+        self.bound_and_clamp(history)
+        # If there are non-state variables, they need to be updated for history:
+        try:
+            # Assuming that node_coupling can have a maximum number of dimensions equal to the state variables,
+            # in the extreme case where all state variables are cvars as well, we set:
+            node_coupling = numpy.zeros((history.shape[0], 1, history.shape[2], 1))
+            for i_time in range(history.shape[1]):
+                self.model.update_non_state_variables(history[:, i_time], node_coupling[:, 0], 0.0)
+            self.bound_and_clamp(history)
+        except:
+            pass
+
+    def update_state(self, state, node_coupling, local_coupling=0.0):
+        # If there are non-state variables, they need to be updated for the initial condition:
+        try:
+            self.model.update_non_state_variables(state, node_coupling, local_coupling)
+            self.bound_and_clamp(state)
+        except:
```

Comment:
bare except is not OK, should prefer to handle this more gracefully or obviate it entirely